java-lib/src/main/java/org/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java

```diff
@@ -0,0 +1,214 @@
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(
+                                inputTypeInfo.getFieldTypes(),
+                                Collections.nCopies(
+                                                outputCols.length,
+                                                ExternalTypeInfo.of(Vector.class))
+                                        .toArray(new TypeInformation[0])),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), outputCols));
+
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) modelDataTable).getTableEnvironment();
+        DataStream<Row> input = tEnv.toDataStream(inputs[0]);
+        DataStream<Tuple2<Integer, Integer>> modelStream =
+                OneHotEncoderModelData.getModelDataStream(modelDataTable);
+
+        GenerateOutputsFunction mapFunction =
+                new GenerateOutputsFunction(getParamMap(), broadcastModelKey);
+
+        Function<List<DataStream<?>>, DataStream<Row>> function =
+                dataStreams -> {
+                    DataStream stream = dataStreams.get(0);
+                    return stream.map(mapFunction, outputTypeInfo);
+                };
+
+        DataStream<Row> output =
+                BroadcastUtils.withBroadcastStream(
+                        Collections.singletonList(input),
+                        Collections.singletonMap(broadcastModelKey, modelStream),
+                        function);
+
+        Table outputTable = tEnv.fromDataStream(output);
+
+        return new Table[] {outputTable};
+    }
+
+    private static class GenerateOutputsFunction extends RichMapFunction<Row, Row>
+            implements OneHotEncoderParams<GenerateOutputsFunction> {
+        private final Map<Param<?>, Object> paramMap;
+        private final String broadcastModelKey;
+        private final String[] inputCols;
+        private Pointer modelPointer = null;
+
+        private GenerateOutputsFunction(Map<Param<?>, Object> paramMap, String broadcastModelKey) {
+            this.paramMap = new HashMap<>(paramMap);
+            this.inputCols = getInputCols();
+            this.broadcastModelKey = broadcastModelKey;
+        }
+
+        @Override
+        public Row map(Row row) throws IOException {
+            if (modelPointer == null) {
+                List<Tuple2<Integer, Integer>> modelDataList =
+                        getRuntimeContext().getBroadcastVariable(broadcastModelKey);
+                modelPointer = loadCppModel(getParamMap(), modelDataList);
+            }
+
+            Row resultRow = new Row(inputCols.length);
+            for (int i = 0; i < inputCols.length; i++) {
+                String inputCol = inputCols[i];
+                int number = ((Number) row.getField(inputCol)).intValue();
+                SparseVectorJNA.ByValue clinkVector =
```

Comment:
It might not be feasible. So far all JNA examples I can found, including the official document, use `ByValue` or `ByReference`. There has not been a direct use of `Structure`s like `SparseVectorJNA`.

Reply:
JNA's javadoc shows that `ByReference` usually has better performance than `ByValue` as it uses direct mapping. I'll change it to `ByReference`.